{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Script to load in a processed seurat object and run pySCENIC\n",
    "\n",
    "To get your processed Seurat object into adata format you can use conversion functions, such as:\n",
    "```r\n",
    "#load custom functions\n",
    "source(\"/pl/active/CSUClinHeme/users/dylan/repos/scrna-seq/analysis-code/customFunctions_Seuratv5.R\")\n",
    "\n",
    "#load in processed R data and subset on cells of interest\n",
    "seu.obj <- readRDS(\"../output/s3/allCells_clean_highRes_integrated.harmony_res1.6_dims45_dist0.1_neigh10_S3.rds\")\n",
    "\n",
    "\n",
    "#stash new idents\n",
    "seu.obj.sub <- subset(seu.obj, subset = clusterID2_integrated.harmony %in% c(3,17,30))\n",
    "\n",
    "cnts <- seu.obj.sub@assays$RNA$counts\n",
    "cnts <- orthogene::convert_orthologs(gene_df = cnts,\n",
    "                                        gene_input = \"rownames\", \n",
    "                                        gene_output = \"rownames\", \n",
    "                                        input_species = \"dog\",\n",
    "                                        output_species = \"human\",\n",
    "                                        non121_strategy = \"drop_both_species\") \n",
    "rownames(cnts) <- unname(rownames(cnts))\n",
    "\n",
    "seu.obj <- CreateSeuratObject(cnts, project = \"humanConvert\", assay = \"RNA\",\n",
    "                                  min.cells = 0, min.features = 0, names.field = 1,\n",
    "                                  names.delim = \"_\", meta.data = seu.obj.sub@meta.data)\n",
    "\n",
    "#covert back to an older Seurat object for compatiability with \n",
    "seu.obj[[\"RNA\"]] <- as(object = seu.obj[[\"RNA\"]], Class = \"Assay\")\n",
    "seu.obj <- NormalizeData(seu.obj)\n",
    "\n",
    "SaveH5Seurat(seu.obj, filename = \"../output/s3/myeloid_branch1_hu.h5Seurat\", verbose = TRUE, overwrite = T)\n",
    "Convert(\"../output/s3/myeloid_branch1_hu.h5Seurat\", dest = \"h5ad\", overwrite = T)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Note: TF files were obtained from https://resources.aertslab.org/cistarget/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import loompy as lp\n",
    "from matplotlib.pyplot import rc_context\n",
    "from MulticoreTSNE import MulticoreTSNE as TSNE\n",
    "from pyscenic.export import export2loom, add_scenic_metadata\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy \n",
    "\n",
    "import glob\n",
    "\n",
    "import json\n",
    "import zlib\n",
    "import base64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "FIGURES_FOLDERNAME=\"../output/scenic/\"\n",
    "def savesvg(fname: str, fig, folder: str=FIGURES_FOLDERNAME) -> None:\n",
    "    \"\"\"\n",
    "    Save figure as vector-based SVG image format.\n",
    "    \"\"\"\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(os.path.join(folder, fname), format='svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2084 × 11606\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.hbm', 'percent.ppbp', 'RNA_snn_res.0.1', 'seurat_clusters', 'int.clusID', 'pANN_0.25_0.01_274', 'DF.classifications_0.25_0.01_274', 'doublet', 'S.Score', 'G2M.Score', 'Phase', 'clusters', 'pANN_0.25_0.19_108', 'DF.classifications_0.25_0.19_108', 'pANN_0.25_0.02_98', 'DF.classifications_0.25_0.02_98', 'pANN_0.25_0.11_115', 'DF.classifications_0.25_0.11_115', 'pANN_0.25_0.005_95', 'DF.classifications_0.25_0.005_95', 'unintegrated_clusters', 'RNA_snn_res.0.6', 'clusterID_integrated.cca', 'clusterID_integrated.harmony', 'clusterID_integrated.joint', 'clusterID_integrated.rcpa', 'name', 'colz', 'cellSource', 'minorIdent', 'majorID', 'major_clusID', 'pseudotime', 'ptime', 'RNA_snn_res.1.6', 'clusterID2_integrated.harmony'\n",
       "    var: 'features'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load in the data\n",
    "adata = sc.read_h5ad(\"../output/s3/myeloid_branch1_hu.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert from numeric to string\n",
    "anno = adata.obs\n",
    "anno['clusterID2_integrated.harmony'] = anno['clusterID2_integrated.harmony'].astype(str)\n",
    "adata.obs = anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert index to gene symbol\n",
    "tempAdata = adata.raw.to_adata()\n",
    "tempAdata.var_names = adata.var['features']\n",
    "adata.raw = tempAdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cant run b/c UMAP coords weren't brought over\n",
    "# sc.pl.umap(adata, color=['IGHM','majorID_sub2','name'], use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create basic row and column attributes for the loom file:\n",
    "row_attrs = {\n",
    "    \"Gene\": np.array(adata.var_names) ,\n",
    "}\n",
    "col_attrs = {\n",
    "    \"CellID\": np.array(adata.obs_names) ,\n",
    "    \"nGene\": np.array( np.sum(adata.X.transpose()>0 , axis=0)).flatten() ,\n",
    "    \"nUMI\": np.array( np.sum(adata.X.transpose() , axis=0)).flatten() ,\n",
    "}\n",
    "lp.create( \"../output/s3/myeloid_branch1_hu.loom\", adata.X.transpose(), row_attrs, col_attrs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The easiest way to run the code is to use a sbatch script, so generate a script with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo \"#!/usr/bin/env bash\" > cute_pyScenic_grn.sbatch\n",
    "!echo \"#SBATCH --job-name=pySCENIC_grn\" >> cute_pyScenic_grn.sbatch\n",
    "\n",
    "!echo \"#SBATCH --nodes=1\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"#SBATCH --ntasks=12\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"#SBATCH --time=00:30:00\" >> cute_pyScenic_grn.sbatch\n",
    "\n",
    "!echo \"#SBATCH --partition=atesting\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"#SBATCH --qos=normal\" >> cute_pyScenic_grn.sbatch\n",
    "\n",
    "!echo \"#SBATCH --mail-type=END\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"#SBATCH --mail-user=dyammons@colostate.edu\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"#SBATCH --output=pySCENIC_grn-%j.log\" >> cute_pyScenic_grn.sbatch\n",
    "\n",
    "!echo \"\" >> cute_pyScenic_grn.sbatch\n",
    "\n",
    "!echo \"module purge\" >> cute_pyScenic_grn.sbatch\n",
    "\n",
    "!echo \"\" >> cute_pyScenic_grn.sbatch\n",
    "\n",
    "!echo \"singularity exec -B $PWD/../ ../software/aertslab-pyscenic-scanpy-0.12.1-1.9.1.sif pyscenic grn \\\\\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"--num_workers \\$SLURM_NTASKS \\\\\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"-o ../output/scenic/adj.csv \\\\\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"../output/s3/myeloid_branch1_hu.loom \\\\\" >> cute_pyScenic_grn.sbatch\n",
    "!echo \"./metaData/allTFs_hg38.txt\" >> cute_pyScenic_grn.sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 4645519\n"
     ]
    }
   ],
   "source": [
    "#submit the job\n",
    "!sbatch cute_pyScenic_grn.sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo \"#!/usr/bin/env bash\" > cute_pyScenic_ctx.sbatch\n",
    "!echo \"#SBATCH --job-name=pySCENIC_ctx\" >> cute_pyScenic_ctx.sbatch\n",
    "\n",
    "!echo \"#SBATCH --nodes=1\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"#SBATCH --ntasks=10\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"#SBATCH --time=00:20:00\" >> cute_pyScenic_ctx.sbatch\n",
    "\n",
    "!echo \"#SBATCH --partition=atesting\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"#SBATCH --qos=normal\" >> cute_pyScenic_ctx.sbatch\n",
    "\n",
    "!echo \"#SBATCH --mail-type=END\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"#SBATCH --mail-user=dyammons@colostate.edu\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"#SBATCH --output=pySCENIC_ctx-%j.log\" >> cute_pyScenic_ctx.sbatch\n",
    "\n",
    "!echo \"\" >> cute_pyScenic_ctx.sbatch\n",
    "\n",
    "!echo \"module purge\" >> cute_pyScenic_ctx.sbatch\n",
    "\n",
    "!echo \"\" >> cute_pyScenic_ctx.sbatch\n",
    "\n",
    "!echo \"singularity exec -B $PWD/../ ../software/aertslab-pyscenic-scanpy-0.12.1-1.9.1.sif pyscenic ctx \\\\\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"../output/scenic/adj.csv \\\\\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"./metaData/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather ./metaData/hg38_500bp_up_100bp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather \\\\\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"--annotations_fname ./metaData/motifs-v10nr_clust-nr.hgnc-m0.001-o0.0.tbl \\\\\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"--expression_mtx_fname ../output/s3/myeloid_branch1_hu.loom \\\\\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"--output ../output/scenic/reg.csv \\\\\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"--mask_dropouts \\\\\" >> cute_pyScenic_ctx.sbatch\n",
    "!echo \"--num_workers \\$SLURM_NTASKS\" >> cute_pyScenic_ctx.sbatch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 4647353\n"
     ]
    }
   ],
   "source": [
    "#submit the job; took 8 min....\n",
    "!sbatch cute_pyScenic_ctx.sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo \"#!/usr/bin/env bash\" > cute_pyScenic_aucell.sbatch\n",
    "!echo \"#SBATCH --job-name=pySCENIC_auc\" >> cute_pyScenic_aucell.sbatch\n",
    "\n",
    "!echo \"#SBATCH --nodes=1\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"#SBATCH --ntasks=4\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"#SBATCH --time=0:10:00\" >> cute_pyScenic_aucell.sbatch\n",
    "\n",
    "!echo \"#SBATCH --partition=atesting\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"#SBATCH --qos=normal\" >> cute_pyScenic_aucell.sbatch\n",
    "\n",
    "!echo \"#SBATCH --mail-type=END\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"#SBATCH --mail-user=dyammons@colostate.edu\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"#SBATCH --output=pySCENIC_auc-%j.log\" >> cute_pyScenic_aucell.sbatch\n",
    "\n",
    "!echo \"\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"module purge\" >> cute_pyScenic_aucell.sbatch\n",
    "\n",
    "!echo \"\" >> cute_pyScenic_aucell.sbatch\n",
    "\n",
    "!echo \"singularity exec -B $PWD/../ ../software/aertslab-pyscenic-scanpy-0.12.1-1.9.1.sif pyscenic aucell \\\\\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"../output/s3/myeloid_branch1_hu.loom \\\\\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"../output/scenic/reg.csv \\\\\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"--output ../output/scenic/myeloid_branch1_hu_output.loom \\\\\" >> cute_pyScenic_aucell.sbatch\n",
    "!echo \"--num_workers \\$SLURM_NTASKS\" >> cute_pyScenic_aucell.sbatch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 4648003\n"
     ]
    }
   ],
   "source": [
    "#took 2 min....\n",
    "!sbatch cute_pyScenic_aucell.sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect SCENIC AUCell output\n",
    "# lf = lp.connect(\"./output/scenic/dc_scenic_output_231023.loom\", mode='r+', validate=False )\n",
    "lf = lp.connect(\"../output/scenic/myeloid_branch1_hu_output.loom\", mode='r+', validate=False )\n",
    "auc_mtx = pd.DataFrame( lf.ca.RegulonsAUC, index=lf.ca.CellID)\n",
    "lf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AHR(+)</th>\n",
       "      <th>ARID3A(+)</th>\n",
       "      <th>ARNT(+)</th>\n",
       "      <th>ATF1(+)</th>\n",
       "      <th>ATF2(+)</th>\n",
       "      <th>ATF3(+)</th>\n",
       "      <th>ATF5(+)</th>\n",
       "      <th>BACH1(+)</th>\n",
       "      <th>BACH2(+)</th>\n",
       "      <th>BCL6(+)</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF768(+)</th>\n",
       "      <th>ZNF770(+)</th>\n",
       "      <th>ZNF774(+)</th>\n",
       "      <th>ZNF782(+)</th>\n",
       "      <th>ZNF784(+)</th>\n",
       "      <th>ZNF792(+)</th>\n",
       "      <th>ZNF8(+)</th>\n",
       "      <th>ZSCAN18(+)</th>\n",
       "      <th>ZSCAN22(+)</th>\n",
       "      <th>ZXDC(+)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BM154801_AAACGAAAGCAACAAT-1</th>\n",
       "      <td>0.006149</td>\n",
       "      <td>0.348276</td>\n",
       "      <td>0.001826</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000840</td>\n",
       "      <td>0.014325</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.030948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010875</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014368</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045211</td>\n",
       "      <td>0.012835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BM154801_AAAGAACAGGTAGACC-1</th>\n",
       "      <td>0.062858</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021552</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010437</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015216</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.038581</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015038</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BM154801_AAAGAACGTAGGTCAG-1</th>\n",
       "      <td>0.004196</td>\n",
       "      <td>0.043966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000818</td>\n",
       "      <td>0.013023</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017666</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.056801</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BM154801_AAAGGGCGTCTGTCCT-1</th>\n",
       "      <td>0.003303</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004158</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005209</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010438</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.025383</td>\n",
       "      <td>0.005090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BM154801_AAAGTCCAGTTGTAGA-1</th>\n",
       "      <td>0.006855</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002738</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008731</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004456</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022126</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C34165634_TTTCGATAGTCTGCAT-1</th>\n",
       "      <td>0.007395</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004604</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018302</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.026293</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.029885</td>\n",
       "      <td>0.004269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C34165634_TTTGATCAGAGAAGGT-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004119</td>\n",
       "      <td>0.001459</td>\n",
       "      <td>0.003173</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018939</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027682</td>\n",
       "      <td>0.003914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C34165634_TTTGGAGCACTCTGCT-1</th>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007062</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005129</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008647</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010632</td>\n",
       "      <td>0.015161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C34165634_TTTGGAGGTAAGGCCA-1</th>\n",
       "      <td>0.006128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004502</td>\n",
       "      <td>0.003360</td>\n",
       "      <td>0.007465</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001164</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005093</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017146</td>\n",
       "      <td>0.004789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C34165634_TTTGTTGCACTTCTCG-1</th>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003161</td>\n",
       "      <td>0.005356</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001638</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012202</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009195</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.021169</td>\n",
       "      <td>0.001806</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2084 rows × 292 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                AHR(+)  ARID3A(+)   ARNT(+)   ATF1(+)  \\\n",
       "BM154801_AAACGAAAGCAACAAT-1   0.006149   0.348276  0.001826  0.000000   \n",
       "BM154801_AAAGAACAGGTAGACC-1   0.062858   0.000000  0.000000  0.021552   \n",
       "BM154801_AAAGAACGTAGGTCAG-1   0.004196   0.043966  0.000000  0.000000   \n",
       "BM154801_AAAGGGCGTCTGTCCT-1   0.003303   0.000000  0.004158  0.000000   \n",
       "BM154801_AAAGTCCAGTTGTAGA-1   0.006855   0.000000  0.002738  0.000000   \n",
       "...                                ...        ...       ...       ...   \n",
       "C34165634_TTTCGATAGTCTGCAT-1  0.007395   0.000000  0.000000  0.000000   \n",
       "C34165634_TTTGATCAGAGAAGGT-1  0.000000   0.000000  0.000000  0.004119   \n",
       "C34165634_TTTGGAGCACTCTGCT-1  0.002098   0.000000  0.000000  0.000000   \n",
       "C34165634_TTTGGAGGTAAGGCCA-1  0.006128   0.000000  0.000000  0.004502   \n",
       "C34165634_TTTGTTGCACTTCTCG-1  0.000083   0.000000  0.000000  0.000000   \n",
       "\n",
       "                               ATF2(+)   ATF3(+)  ATF5(+)  BACH1(+)  BACH2(+)  \\\n",
       "BM154801_AAACGAAAGCAACAAT-1   0.000840  0.014325      0.0  0.030948       0.0   \n",
       "BM154801_AAAGAACAGGTAGACC-1   0.000000  0.010437      0.0  0.015216       0.0   \n",
       "BM154801_AAAGAACGTAGGTCAG-1   0.000818  0.013023      0.0  0.000000       0.0   \n",
       "BM154801_AAAGGGCGTCTGTCCT-1   0.000000  0.005209      0.0  0.000000       0.0   \n",
       "BM154801_AAAGTCCAGTTGTAGA-1   0.000000  0.008731      0.0  0.008060       0.0   \n",
       "...                                ...       ...      ...       ...       ...   \n",
       "C34165634_TTTCGATAGTCTGCAT-1  0.000000  0.004604      0.0  0.000000       0.0   \n",
       "C34165634_TTTGATCAGAGAAGGT-1  0.001459  0.003173      0.0  0.000259       0.0   \n",
       "C34165634_TTTGGAGCACTCTGCT-1  0.000000  0.007062      0.0  0.005129       0.0   \n",
       "C34165634_TTTGGAGGTAAGGCCA-1  0.003360  0.007465      0.0  0.001164       0.0   \n",
       "C34165634_TTTGTTGCACTTCTCG-1  0.003161  0.005356      0.0  0.001638       0.0   \n",
       "\n",
       "                               BCL6(+)  ...  ZNF768(+)  ZNF770(+)  ZNF774(+)  \\\n",
       "BM154801_AAACGAAAGCAACAAT-1   0.010875  ...   0.014368        0.0        0.0   \n",
       "BM154801_AAAGAACAGGTAGACC-1   0.038581  ...   0.000000        0.0        0.0   \n",
       "BM154801_AAAGAACGTAGGTCAG-1   0.017666  ...   0.000000        0.0        0.0   \n",
       "BM154801_AAAGGGCGTCTGTCCT-1   0.010438  ...   0.000000        0.0        0.0   \n",
       "BM154801_AAAGTCCAGTTGTAGA-1   0.004456  ...   0.000000        0.0        0.0   \n",
       "...                                ...  ...        ...        ...        ...   \n",
       "C34165634_TTTCGATAGTCTGCAT-1  0.018302  ...   0.000000        0.0        0.0   \n",
       "C34165634_TTTGATCAGAGAAGGT-1  0.018939  ...   0.000000        0.0        0.0   \n",
       "C34165634_TTTGGAGCACTCTGCT-1  0.008647  ...   0.000000        0.0        0.0   \n",
       "C34165634_TTTGGAGGTAAGGCCA-1  0.005093  ...   0.000000        0.0        0.0   \n",
       "C34165634_TTTGTTGCACTTCTCG-1  0.012202  ...   0.009195        0.0        0.0   \n",
       "\n",
       "                              ZNF782(+)  ZNF784(+)  ZNF792(+)  ZNF8(+)  \\\n",
       "BM154801_AAACGAAAGCAACAAT-1    0.000000        0.0        0.0      0.0   \n",
       "BM154801_AAAGAACAGGTAGACC-1    0.000000        0.0        0.0      0.0   \n",
       "BM154801_AAAGAACGTAGGTCAG-1    0.000000        0.0        0.0      0.0   \n",
       "BM154801_AAAGGGCGTCTGTCCT-1    0.000000        0.0        0.0      0.0   \n",
       "BM154801_AAAGTCCAGTTGTAGA-1    0.000000        0.0        0.0      0.0   \n",
       "...                                 ...        ...        ...      ...   \n",
       "C34165634_TTTCGATAGTCTGCAT-1   0.026293        0.0        0.0      0.0   \n",
       "C34165634_TTTGATCAGAGAAGGT-1   0.000000        0.0        0.0      0.0   \n",
       "C34165634_TTTGGAGCACTCTGCT-1   0.000000        0.0        0.0      0.0   \n",
       "C34165634_TTTGGAGGTAAGGCCA-1   0.000000        0.0        0.0      0.0   \n",
       "C34165634_TTTGTTGCACTTCTCG-1   0.003707        0.0        0.0      0.0   \n",
       "\n",
       "                              ZSCAN18(+)  ZSCAN22(+)   ZXDC(+)  \n",
       "BM154801_AAACGAAAGCAACAAT-1          0.0    0.045211  0.012835  \n",
       "BM154801_AAAGAACAGGTAGACC-1          0.0    0.015038  0.000000  \n",
       "BM154801_AAAGAACGTAGGTCAG-1          0.0    0.056801  0.000000  \n",
       "BM154801_AAAGGGCGTCTGTCCT-1          0.0    0.025383  0.005090  \n",
       "BM154801_AAAGTCCAGTTGTAGA-1          0.0    0.022126  0.000000  \n",
       "...                                  ...         ...       ...  \n",
       "C34165634_TTTCGATAGTCTGCAT-1         0.0    0.029885  0.004269  \n",
       "C34165634_TTTGATCAGAGAAGGT-1         0.0    0.027682  0.003914  \n",
       "C34165634_TTTGGAGCACTCTGCT-1         0.0    0.010632  0.015161  \n",
       "C34165634_TTTGGAGGTAAGGCCA-1         0.0    0.017146  0.004789  \n",
       "C34165634_TTTGTTGCACTTCTCG-1         0.0    0.021169  0.001806  \n",
       "\n",
       "[2084 rows x 292 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auc_mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyscenic.rss import regulon_specificity_scores\n",
    "from pyscenic.plotting import plot_rss\n",
    "import matplotlib.pyplot as plt\n",
    "# from adjustText import adjust_text\n",
    "import seaborn as sns\n",
    "from pyscenic.binarization import binarize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "cellAnnot = pd.concat(\n",
    "    [\n",
    "        pd.DataFrame( adata.obs.index.tolist(), index=adata.obs['clusterID2_integrated.harmony'].tolist() )\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "cellAnnot.columns = [\n",
    " 'clusterID2_integrated.harmony']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AHR(+)</th>\n",
       "      <th>ARID3A(+)</th>\n",
       "      <th>ARNT(+)</th>\n",
       "      <th>ATF1(+)</th>\n",
       "      <th>ATF2(+)</th>\n",
       "      <th>ATF3(+)</th>\n",
       "      <th>ATF5(+)</th>\n",
       "      <th>BACH1(+)</th>\n",
       "      <th>BACH2(+)</th>\n",
       "      <th>BCL6(+)</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF768(+)</th>\n",
       "      <th>ZNF770(+)</th>\n",
       "      <th>ZNF774(+)</th>\n",
       "      <th>ZNF782(+)</th>\n",
       "      <th>ZNF784(+)</th>\n",
       "      <th>ZNF792(+)</th>\n",
       "      <th>ZNF8(+)</th>\n",
       "      <th>ZSCAN18(+)</th>\n",
       "      <th>ZSCAN22(+)</th>\n",
       "      <th>ZXDC(+)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.345594</td>\n",
       "      <td>0.531878</td>\n",
       "      <td>0.278413</td>\n",
       "      <td>0.255813</td>\n",
       "      <td>0.276130</td>\n",
       "      <td>0.361036</td>\n",
       "      <td>0.182742</td>\n",
       "      <td>0.426544</td>\n",
       "      <td>0.194784</td>\n",
       "      <td>0.382647</td>\n",
       "      <td>...</td>\n",
       "      <td>0.275423</td>\n",
       "      <td>0.191327</td>\n",
       "      <td>0.173165</td>\n",
       "      <td>0.212178</td>\n",
       "      <td>0.187939</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.197946</td>\n",
       "      <td>0.275173</td>\n",
       "      <td>0.385119</td>\n",
       "      <td>0.336470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0.431298</td>\n",
       "      <td>0.184048</td>\n",
       "      <td>0.257456</td>\n",
       "      <td>0.254651</td>\n",
       "      <td>0.268700</td>\n",
       "      <td>0.285789</td>\n",
       "      <td>0.176422</td>\n",
       "      <td>0.333909</td>\n",
       "      <td>0.206036</td>\n",
       "      <td>0.300222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.212837</td>\n",
       "      <td>0.175888</td>\n",
       "      <td>0.167445</td>\n",
       "      <td>0.181953</td>\n",
       "      <td>0.191452</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.216853</td>\n",
       "      <td>0.212957</td>\n",
       "      <td>0.314162</td>\n",
       "      <td>0.245767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.401027</td>\n",
       "      <td>0.244819</td>\n",
       "      <td>0.250585</td>\n",
       "      <td>0.283455</td>\n",
       "      <td>0.410835</td>\n",
       "      <td>0.558180</td>\n",
       "      <td>0.182847</td>\n",
       "      <td>0.340268</td>\n",
       "      <td>0.187611</td>\n",
       "      <td>0.567580</td>\n",
       "      <td>...</td>\n",
       "      <td>0.341794</td>\n",
       "      <td>0.172368</td>\n",
       "      <td>0.182956</td>\n",
       "      <td>0.301025</td>\n",
       "      <td>0.185764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.224002</td>\n",
       "      <td>0.235091</td>\n",
       "      <td>0.458189</td>\n",
       "      <td>0.534193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 292 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      AHR(+)  ARID3A(+)   ARNT(+)   ATF1(+)   ATF2(+)   ATF3(+)   ATF5(+)  \\\n",
       "17  0.345594   0.531878  0.278413  0.255813  0.276130  0.361036  0.182742   \n",
       "30  0.431298   0.184048  0.257456  0.254651  0.268700  0.285789  0.176422   \n",
       "3   0.401027   0.244819  0.250585  0.283455  0.410835  0.558180  0.182847   \n",
       "\n",
       "    BACH1(+)  BACH2(+)   BCL6(+)  ...  ZNF768(+)  ZNF770(+)  ZNF774(+)  \\\n",
       "17  0.426544  0.194784  0.382647  ...   0.275423   0.191327   0.173165   \n",
       "30  0.333909  0.206036  0.300222  ...   0.212837   0.175888   0.167445   \n",
       "3   0.340268  0.187611  0.567580  ...   0.341794   0.172368   0.182956   \n",
       "\n",
       "    ZNF782(+)  ZNF784(+)  ZNF792(+)   ZNF8(+)  ZSCAN18(+)  ZSCAN22(+)  \\\n",
       "17   0.212178   0.187939        NaN  0.197946    0.275173    0.385119   \n",
       "30   0.181953   0.191452        NaN  0.216853    0.212957    0.314162   \n",
       "3    0.301025   0.185764        NaN  0.224002    0.235091    0.458189   \n",
       "\n",
       "     ZXDC(+)  \n",
       "17  0.336470  \n",
       "30  0.245767  \n",
       "3   0.534193  \n",
       "\n",
       "[3 rows x 292 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rss = regulon_specificity_scores(auc_mtx, adata.obs['clusterID2_integrated.harmony'])\n",
    "rss.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['AHR(+)', 'ARID3A(+)', 'ARNT(+)', 'ATF1(+)', 'ATF2(+)', 'ATF3(+)',\n",
       "       'ATF5(+)', 'BACH1(+)', 'BACH2(+)', 'BCL6(+)', 'BHLHE40(+)',\n",
       "       'BRCA1(+)', 'CBFB(+)', 'CCNT2(+)', 'CDC5L(+)', 'CEBPB(+)',\n",
       "       'CEBPE(+)', 'CEBPG(+)', 'CHD2(+)', 'CLOCK(+)', 'CNOT3(+)',\n",
       "       'CREB1(+)', 'CUX1(+)', 'DBP(+)', 'DDIT3(+)', 'DEAF1(+)', 'E2F1(+)',\n",
       "       'E2F2(+)', 'E2F3(+)', 'E2F6(+)', 'E2F7(+)', 'E2F8(+)', 'EBF1(+)',\n",
       "       'ELF1(+)', 'ELF2(+)', 'ELF4(+)', 'ELK1(+)', 'ELK3(+)', 'ELK4(+)',\n",
       "       'EP300(+)', 'ERF(+)', 'ERG(+)', 'ESRRA(+)', 'ETS1(+)', 'ETS2(+)',\n",
       "       'ETV3(+)', 'ETV5(+)', 'ETV6(+)', 'ETV7(+)', 'FLI1(+)', 'FOS(+)',\n",
       "       'FOSB(+)', 'FOSL1(+)', 'FOSL2(+)', 'FOXJ2(+)', 'FOXK1(+)',\n",
       "       'FOXM1(+)', 'FOXN2(+)', 'FOXO1(+)', 'FOXO4(+)', 'GABPA(+)',\n",
       "       'GABPB1(+)', 'GATA1(+)', 'GATA2(+)', 'GATA3(+)', 'GFI1(+)',\n",
       "       'GFI1B(+)', 'GLIS1(+)', 'GMEB1(+)', 'GTF2IRD1(+)', 'GTF3A(+)',\n",
       "       'HDAC2(+)', 'HIF1A(+)', 'HLTF(+)', 'HMG20B(+)', 'HNF4A(+)',\n",
       "       'HOXA10(+)', 'HOXA4(+)', 'HOXA6(+)', 'HOXA9(+)', 'HOXB2(+)',\n",
       "       'HOXB4(+)', 'HOXB5(+)', 'HOXC4(+)', 'HSF1(+)', 'HSF2(+)',\n",
       "       'HTATIP2(+)', 'IKZF1(+)', 'IKZF2(+)', 'IKZF3(+)', 'IKZF5(+)',\n",
       "       'ILF2(+)', 'IRF1(+)', 'IRF3(+)', 'IRF4(+)', 'IRF8(+)', 'JUNB(+)',\n",
       "       'JUND(+)', 'KLF12(+)', 'KLF3(+)', 'KLF4(+)', 'KLF6(+)', 'LEF1(+)',\n",
       "       'LHX6(+)', 'LTF(+)', 'MAF(+)', 'MAFB(+)', 'MAFK(+)', 'MAX(+)',\n",
       "       'MAZ(+)', 'MBD2(+)', 'MECP2(+)', 'MED30(+)', 'MEIS1(+)', 'MGA(+)',\n",
       "       'MITF(+)', 'MLX(+)', 'MNT(+)', 'MXD1(+)', 'MXD3(+)', 'MXD4(+)',\n",
       "       'MXI1(+)', 'MYB(+)', 'MYBL1(+)', 'MYBL2(+)', 'MYC(+)', 'MZF1(+)',\n",
       "       'NELFB(+)', 'NFATC1(+)', 'NFATC2(+)', 'NFATC3(+)', 'NFE2(+)',\n",
       "       'NFE2L1(+)', 'NFE2L2(+)', 'NFE2L3(+)', 'NFIA(+)', 'NFIC(+)',\n",
       "       'NFIX(+)', 'NFKB1(+)', 'NFKB2(+)', 'NFYB(+)', 'NFYC(+)',\n",
       "       'NKX3-1(+)', 'NR1I3(+)', 'NR2C2(+)', 'NR3C1(+)', 'NR3C2(+)',\n",
       "       'NR4A1(+)', 'NR4A2(+)', 'NRF1(+)', 'PAX5(+)', 'PBX1(+)', 'PBX3(+)',\n",
       "       'PLAGL2(+)', 'POU2F1(+)', 'PPARG(+)', 'PRDM10(+)', 'PRDM5(+)',\n",
       "       'PURA(+)', 'RARA(+)', 'RARG(+)', 'RBPJ(+)', 'REL(+)', 'RELB(+)',\n",
       "       'REST(+)', 'RFX1(+)', 'RFX2(+)', 'RFX3(+)', 'RFX5(+)', 'RFX7(+)',\n",
       "       'RFXANK(+)', 'RREB1(+)', 'RUNX1(+)', 'RUNX3(+)', 'RXRA(+)',\n",
       "       'RXRB(+)', 'SCAND1(+)', 'SF1(+)', 'SIN3A(+)', 'SIRT6(+)',\n",
       "       'SMAD1(+)', 'SMAD3(+)', 'SOX12(+)', 'SOX13(+)', 'SOX4(+)',\n",
       "       'SOX5(+)', 'SOX6(+)', 'SOX8(+)', 'SP1(+)', 'SP2(+)', 'SP3(+)',\n",
       "       'SP4(+)', 'SPI1(+)', 'SREBF1(+)', 'SRF(+)', 'STAT1(+)', 'STAT2(+)',\n",
       "       'STAT3(+)', 'STAT5A(+)', 'STAT5B(+)', 'STAT6(+)', 'TAL1(+)',\n",
       "       'TBP(+)', 'TCF12(+)', 'TCF3(+)', 'TCF4(+)', 'TCF7(+)', 'TCF7L1(+)',\n",
       "       'TCF7L2(+)', 'TEF(+)', 'TERF1(+)', 'TFAP4(+)', 'TFDP1(+)',\n",
       "       'TFDP2(+)', 'TFE3(+)', 'TFEC(+)', 'TGIF1(+)', 'TGIF2(+)',\n",
       "       'THRB(+)', 'TOPORS(+)', 'USF1(+)', 'USF2(+)', 'VDR(+)', 'YY1(+)',\n",
       "       'ZBED1(+)', 'ZBTB11(+)', 'ZBTB12(+)', 'ZBTB14(+)', 'ZBTB16(+)',\n",
       "       'ZBTB18(+)', 'ZBTB2(+)', 'ZBTB25(+)', 'ZBTB7A(+)', 'ZBTB7B(+)',\n",
       "       'ZEB1(+)', 'ZFHX2(+)', 'ZFHX3(+)', 'ZFP14(+)', 'ZFP64(+)',\n",
       "       'ZFP82(+)', 'ZHX3(+)', 'ZNF112(+)', 'ZNF131(+)', 'ZNF143(+)',\n",
       "       'ZNF148(+)', 'ZNF184(+)', 'ZNF200(+)', 'ZNF202(+)', 'ZNF214(+)',\n",
       "       'ZNF219(+)', 'ZNF23(+)', 'ZNF24(+)', 'ZNF250(+)', 'ZNF274(+)',\n",
       "       'ZNF287(+)', 'ZNF296(+)', 'ZNF304(+)', 'ZNF311(+)', 'ZNF316(+)',\n",
       "       'ZNF32(+)', 'ZNF333(+)', 'ZNF34(+)', 'ZNF367(+)', 'ZNF407(+)',\n",
       "       'ZNF410(+)', 'ZNF467(+)', 'ZNF484(+)', 'ZNF550(+)', 'ZNF575(+)',\n",
       "       'ZNF579(+)', 'ZNF580(+)', 'ZNF584(+)', 'ZNF605(+)', 'ZNF624(+)',\n",
       "       'ZNF639(+)', 'ZNF652(+)', 'ZNF654(+)', 'ZNF691(+)', 'ZNF692(+)',\n",
       "       'ZNF7(+)', 'ZNF740(+)', 'ZNF76(+)', 'ZNF768(+)', 'ZNF770(+)',\n",
       "       'ZNF774(+)', 'ZNF782(+)', 'ZNF784(+)', 'ZNF792(+)', 'ZNF8(+)',\n",
       "       'ZSCAN18(+)', 'ZSCAN22(+)', 'ZXDC(+)'], dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rss.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "rss = rss.rename(index={'0' : '17',\n",
    "                        '1' : '30',\n",
    "                        '2' : '3'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame.to_csv(rss, \"../output/scenic/myeloid_branch1_hu_rss.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "```r\n",
    "library(tidyverse)\n",
    "library(ggrepel)\n",
    "library(patchwork)\n",
    "\n",
    "df <- read.csv(\"../output/scenic/myeloid_branch1_hu_rss.csv\", row.names = 1)\n",
    "colnames(df) <- gsub('\\\\.\\\\.\\\\.',\"\",colnames(df))\n",
    "df.all <- t(df) %>% as.data.frame() %>% rownames_to_column() %>% na.omit()\n",
    "\n",
    "labCut = 10\n",
    "pi <- lapply(2:ncol(df.all), function(x){\n",
    "    #prep data\n",
    "    celltype <- colnames(df.all)[x]\n",
    "    df <- df.all[ ,c(\"rowname\",celltype)] \n",
    "    colnames(df) <- c(\"gene\", \"clus\")\n",
    "    df <- df %>% arrange(desc(clus))\n",
    "    df[ ,\"gene\"] <- factor(df[ ,\"gene\"], levels = df[ ,\"gene\"])\n",
    "    df <- df %>% mutate(colz = ifelse(row_number() < labCut+1, \"#FF6961\", \"lightblue\"),\n",
    "                        labz = ifelse(row_number() < labCut+1, as.character(gene), NA))\n",
    "    \n",
    "    #plot the data\n",
    "    p <- ggplot(df, aes(x = gene, y = clus, label = labz)) +\n",
    "    geom_point(colour = df$colz, size = 2) + \n",
    "    geom_text_repel(\n",
    "        force = 0.01,\n",
    "        nudge_x = nrow(df)*0.5,\n",
    "        direction = \"y\",\n",
    "        seed = 42, \n",
    "        box.padding = 0.25,\n",
    "        #hjust = 0,\n",
    "        #segment.size = 0.2, \n",
    "        size = 4, \n",
    "        color = df$colz,\n",
    "        max.iter = 100000000,\n",
    "        max.overlaps = 10\n",
    "    ) + \n",
    "    labs(\n",
    "        title = celltype,\n",
    "        x = \"Regulon\",\n",
    "        y = \"rss\"\n",
    "    ) + \n",
    "    theme_classic() + \n",
    "    theme(\n",
    "        axis.ticks.x = element_blank(),\n",
    "        axis.text.x = element_blank(),\n",
    "        panel.border = element_blank(),\n",
    "        plot.margin = unit(c(10, 10, 10, 10), \"pt\"),\n",
    "        panel.background = element_rect(fill = \"transparent\",colour = NA),\n",
    "        plot.background = element_rect(fill = \"transparent\",colour = NA),\n",
    "        panel.grid.major = element_blank(), \n",
    "        panel.grid.minor = element_blank()\n",
    "    ) + \n",
    "    scale_y_continuous(\n",
    "        breaks = function(y) {\n",
    "            seq(floor(min(y, digits = 1)), \n",
    "            ceiling(max(y, digits = 1)), \n",
    "            by = 0.1)\n",
    "        }\n",
    "    ) + coord_cartesian(expand = TRUE, clip = \"off\")\n",
    "\n",
    "    #return/save plot\n",
    "    ggsave(paste0(\"../output/scenic/\", gsub(\" \", \"_\", celltype), \"_rss.png\"), height = 5, width = 3)\n",
    "    return(p)\n",
    "})\n",
    "\n",
    "\n",
    "p <- Reduce( `+`, pi) + plot_layout(ncol = 3)\n",
    "ggsave(plot = p, \"../output/scenic/myeloid_branch1_rss.png\", height = 4, width = 6)\n",
    "\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scPy",
   "language": "python",
   "name": "scpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
